src/assemble_workflow/dist.py

```diff
@@ -36,42 +41,145 @@ def extract(self, dest):
 
         raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), os.path.join(dest, "*"))
 
-    def build(self, name, dest):
-        self.__build__(name, dest)
+    def build(self, bundle_recorder, dest):
+        name = bundle_recorder.package_name
+        self.__build__(bundle_recorder, dest)
         path = os.path.join(dest, name)
         shutil.copyfile(name, path)
         logging.info(f"Published {path}.")
 
     @classmethod
-    def from_path(cls, name, path):
-        ext = os.path.splitext(path)[1]
-        if ext == ".gz":
-            return DistTar(name, path)
-        elif ext == ".zip":
-            return DistZip(name, path)
+    def create_dist(cls, name, path, distribution):
+        DISTRIBUTIONS = {
+            "rpm": DistRpm(name, path),
+            "tar": DistTar(name, path),
+            "zip": DistZip(name, path),
+        }
+        if distribution not in DISTRIBUTIONS:
+            raise ValueError("Distribution not specified or invalid distribution")
         else:
-            raise ValueError(f'Invalid min "dist" extension in input artifacts: {ext} ({path}).')
+            return DISTRIBUTIONS[distribution]
 
 
 class DistZip(Dist):
     def __extract__(self, dest):
         with ZipFile(self.path, "r") as zip:
             zip.extractall(dest)
 
-    def __build__(self, name, dest):
-        with ZipFile(name, "w", zipfile.ZIP_DEFLATED) as zip:
+    def __build__(self, bundle_recorder, dest):
+        with ZipFile(bundle_recorder.package_name, "w", zipfile.ZIP_DEFLATED) as zip:
             rootlen = len(self.archive_path) + 1
             for base, dirs, files in os.walk(self.archive_path):
                 for file in files:
                     fn = os.path.join(base, file)
                     zip.write(fn, fn[rootlen:])
 
+    @property
+    def get_distribution(self):
+        return "zip"
+
 
 class DistTar(Dist):
     def __extract__(self, dest):
         with tarfile.open(self.path, "r:gz") as tar:
             tar.extractall(dest)
 
-    def __build__(self, name, dest):
-        with tarfile.open(name, "w:gz") as tar:
+    def __build__(self, bundle_recorder, dest):
+        with tarfile.open(bundle_recorder.package_name, "w:gz") as tar:
             tar.add(self.archive_path, arcname=os.path.basename(self.archive_path))
+
+    @property
+    def get_distribution(self):
+        return "tar"
+
+
+class DistRpm(Dist):
```

Comment:
Given how long these classes are becoming, move them to their own files (dist_rpm.py, etc.), and have their own `test_...` files.

Then `create_dist` will have to go into a new class `Dists` (plural) that will have a `create` (just call it that) static method.

Reply:
Thank you for your comments! @dblock We are thinking about extracting the whole fpm build process of DistRpm() into its parent class Dist() as a method called fpm_build() or something else, since we are going to add one more DistDeb() which shares the same build process as DistRpm() in the future. Then, these two subclasses could inherit fpm_build() and we only need to implement this method once to increase reusability of the code. Do you think this idea make sense?

Reply:
FPM is its own tool/process, so you likely will want a utility class called `FpmBuilder` or something like that.

Reply:
Ok. That makes sense